flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -508,6 +510,8 @@ def fir_StoreOp : fir_Op<"store", []> {
   let verifier = [{
     if (value().getType() != fir::dyn_cast_ptrEleTy(memref().getType()))
       return emitOpError("store value type must match memory reference type");
+    if (fir::isa_box_type(value().getType()))
+      return emitOpError("!fir.box cannot be stored");
```

Comment:
This is a TODO though, right? We have no other way to initialize a box in memory otherwise. (I'm ignoring calls to opaque operations in the library intentionally.)

Reply:
It most likely is. Right now, allowing this without a change in codegen crashes because we need to do a memcopy instead of a LLVM::StoreOp for boxes (since both the "value" and address are actually both addresses).

It is not 100% clear to me how we want to make `fir.ref<fir.box>` writable in fir (to avoid always relying on opaque runtime calls). I see serval options (there may be more):
- option 1: creating a complete new `fir.box` using `fir.embox`, and implement `fir.store fir.ref<box> fir.box`
- option 2: having fir operations to write to `fir.ref<fir.box>` fields (like  `fir.store_box_addr`, `fir.store_box_dims`,...)
- option 3: casting `fir.ref<fir.box<>> to a `fir.ref<fir.array<?x i8>>` and addressing the fields in lowering assuming what the descriptor layout is.

option 1 advantages: more control over the state of descriptors, pretty much doable right now with a few modification in codegen around fir.store. No descriptor layout knowledge needed in lowering. 
option 1 drawbacks: I think we would leave LLVM having to optimized out a lot of useless things (a temporary descriptor, and we would emit write to all fields when we may actually want to only write a few).

option 2 advantages: No temp descriptor created, only emitting writes to the fields we need to writes. No descriptor layout knowledge needed in lowering. 
option 2 drawbacks: New ops needed in fir. More risk to get the descriptor in an invalid state (e.g some part of lowering code assume a field was already set, but that is not always the case...).

option 3 advantages: doable right now without any change in codegen
option 3 drawbacks: Layout of descriptors needs to be known in the bridge. More opaque fir to read (hard to understand which fields is being written). Same invalid descriptor state risk as in 2.


Option 2 would probably be my favorite, but I need to check if we could use this in the context of global descriptor initializer (which is the main place we absolutely want to initialize some fields without runtime. Right now, option 3 might be the only option there).

Reply:
(3) does not appeal to me. I don't want to fully expose the internal organs of how a box is ultimately lowered at the FIR level. For instance, if a new layout should come around, that would mean reworking everything from FIR to the end. If several layouts are supported by different backends, the complexity would be magnified. Therefore, I want a box to be considered an indivisible abstract value.

The ability to build a box is currently atomic and done by a single instruction. Those are strengths of the IR. As you point out, complicating the reasoning about correctness when all or part of a box is left in an undefined state would be a real disadvantage.